### PR TITLE
Type fixes and logic extraction

### PR DESCRIPTION
**Pull request by kristojorg**
Hi all,

This PR provides some type fixes that I needed in order to work with the repo properly in circulation-patron-web. It does two other things as well:

1. I created a hook called `usePathFor` which is a simple wrapper around `React.useContext(PathForContext)`, providing additional null checks that are useful so that the context consumer doesn't have to check for undefined every time they want to use the context. 
2. I extracted a bit of the logic in `DownloadButton` so that I could import and reuse it in circulation-patron-web. 

Let me know if you have any questions

### CODE REVIEW

**Comment by adriana-alter** [comment]
Why are so many of the properties optional now?

**Comment by kristojorg** [comment]
In circulation-patron-web, as well as in OPDS-web-client, typescripts `strictNullChecks` is not turned on, which means that even when a property is required in a function call, typescript is only checking that something is passed in, not that that thing is defined. 

In practice, many of these functions without optional properties were being used with undefined or null being passed in. In circulation-patron-web, I decided to turn on `strictNullChecks`, which meant I was not able to use them in the same way without making the properties optional. Turning this setting on has the significant benefit of guaranteeing that you will never get an undefined property access type error in runtime. 

I would suggest it is turned on in opds-web-client as well, but don't have time to fix the resulting errors. 

Long story short, these properties that are now optional were previously effectively optional because typescript wasn't checking whether the passed in variable was null/undefined.

File: packages/opds-web-client/src/state.ts [hunk 1]
```diff
@@ -18,8 +18,8 @@ export interface State {
     package, but it's available for other apps if they need to build the state
     for server-side rendering. */
 export default function buildInitialState(
-  collectionUrl: string,
-  bookUrl: string
+  collectionUrl?: string,
+  bookUrl?: string
```
Comment (kristojorg) [inline packages/opds-web-client/src/state.ts]: For example, @adriana-alter , here you can imagine that the initial state could be built without a bookUrl - in fact unless the user is specifically requesting a /book/:bookUrl route, I believe there would be no bookUrl provided here, but it is marked as required. Which means that when you turn on strictNullChecks, I had to pass in a string at all times to this function.

**Comment by kristojorg** [comment]
Here is some more background on that compiler flag: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html

For me, checking that variables are defined is possibly the single most useful thing that typescript can do

File: packages/opds-web-client/src/components/DownloadButton.tsx [hunk 2]
```diff
@@ -75,7 +76,7 @@ export default class DownloadButton extends React.Component<
       return this.props.fulfill(this.props.url).then(blob => {
         download(
           blob,
-          this.generateFilename(this.props.title),
+          generateFilename(this.props.title, this.fileExtension()),
```
Comment (EdwinGuzman) [inline packages/opds-web-client/src/components/DownloadButton.tsx]: This is fine but the reference to the original `generateFilename` function is still in this file and can be removed.

File: packages/opds-web-client/src/components/DownloadButton.tsx [hunk 3]
```diff
@@ -124,9 +118,8 @@ export default class DownloadButton extends React.Component<
     ) {
       return "Read Online";
     }
-    let type = this.fileExtension()
-      .replace(".", "")
-      .toUpperCase();
+    let type = typeMap[this.mimeType()]?.name;
```
Comment (EdwinGuzman) [inline packages/opds-web-client/src/components/DownloadButton.tsx]: Nice.

File: packages/opds-web-client/src/components/DownloadButton.tsx [hunk 4]
```diff
@@ -107,14 +108,7 @@ export default class DownloadButton extends React.Component<
...
   fileExtension() {
-    return (
-      {
-        "application/epub+zip": ".epub",
-        "application/pdf": ".pdf",
-        "application/vnd.adobe.adept+xml": ".acsm",
-        "application/x-mobipocket-ebook": ".mobi"
-      }[this.mimeType()] || ""
-    );
+    return typeMap[this.mimeType()]?.extension ?? "";
```
Comment (EdwinGuzman) [inline packages/opds-web-client/src/components/DownloadButton.tsx]: So, basically, if the left side is ever null or undefined, then and only then return `""`, right?

Comment (EdwinGuzman) [reply packages/opds-web-client/src/components/DownloadButton.tsx]: For Typescript syntax that is new and/or is not already in the repo, do you mind writing a small comment about it? It'll make it easier to read.

File: packages/opds-web-client/src/components/__tests__/DownloadButton-test.tsx [hunk 5]
```diff
@@ -43,15 +43,15 @@ describe("DownloadButton", () => {
   it("shows button", () => {
     let button = wrapper.find("button");
     expect(button.props().style).to.deep.equal(style);
-    expect(button.text()).to.equal("Download EPUB");
+    expect(button.text()).to.equal("Download ePub");
```
Comment (EdwinGuzman) [inline packages/opds-web-client/src/components/__tests__/DownloadButton-test.tsx]: I forgot if we talked about this before, but is it okay to switch to `"ePub"` now?

File: packages/opds-web-client/src/components/context/StoreContext.tsx [hunk 6]
```diff
@@ -1,16 +1,16 @@
...
 import * as Redux from "redux";
 import { State } from "../../state";
 import AuthPlugin from "../../AuthPlugin";
 import buildStore from "../../store";
 import { PathForContext } from "./PathForContext";
 import BasicAuthPlugin from "../../BasicAuthPlugin";
 
-type OPDSStoreProps = {
-  children: React.ReactElement;
+export type OPDSStoreProps = {
```
Comment (EdwinGuzman) [inline packages/opds-web-client/src/components/context/StoreContext.tsx]: Was this exported for cpw?

File: packages/opds-web-client/src/utils/file.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    extension: ".acsm",
+    name: "ACSM"
+  },
+  "application/x-mobipocket-ebook": {
+    extension: ".mobi",
+    name: "MOBI"
+  }
+};
+
+export const generateFilename = (str: string, extension: string) => {
```
Comment (EdwinGuzman) [inline packages/opds-web-client/src/utils/file.ts]: Can you write tests for this function now that it's a util function?
When you remove the `generateFilename` method from the DownloadButton component, a test should fail.

File: packages/opds-web-client/src/state.ts (hunk 1 above)
Comment (adriana-alter) [reply packages/opds-web-client/src/state.ts]: Makes sense; thank you!

File: packages/opds-web-client/src/components/__tests__/DownloadButton-test.tsx (hunk 5 above)
Comment (kristojorg) [reply packages/opds-web-client/src/components/__tests__/DownloadButton-test.tsx]: Ah right, I meant to ask this question in my PR - I'm not sure if it okay or not. We can keep EPUB, I just see ePub around the internet most of the time

File: packages/opds-web-client/src/components/context/StoreContext.tsx (hunk 6 above)
Comment (kristojorg) [reply packages/opds-web-client/src/components/context/StoreContext.tsx]: Yep

File: packages/opds-web-client/src/utils/file.ts (hunk 7 above)
Comment (kristojorg) [reply packages/opds-web-client/src/utils/file.ts]: Yep on it

File: packages/opds-web-client/src/components/DownloadButton.tsx (hunk 4 above)
Comment (kristojorg) [reply packages/opds-web-client/src/components/DownloadButton.tsx]: Will do, and yes that's right. It's very similar to doing ` const blah = maybeDefined || fallback`, except that it only uses the fallback for `null` and `undefined`, while ` || ` will also fall back for falsy values (0, false, etc).

File: packages/opds-web-client/src/utils/file.ts (hunk 7 above)
Comment (kristojorg) [reply packages/opds-web-client/src/utils/file.ts]: I didn't see any existing tests but I did write some of my own. Let me know if you're sufficient

**Comment by kristojorg** [comment]
I also now have made a string literal type for the `OpenAccessLinksType` which is used to make sure that the `typeMap` for example is complete. It refines the type from string, to only a few strings that are possible. 

We should do the same for `fullfillmentLinks`, but I haven't gotten around to dealing with those. They should be the same I believe though, correct? If so, I'll use the same enum there and we can get that change into this PR.

File: packages/opds-web-client/src/components/__tests__/DownloadButton-test.tsx (hunk 5 above)
Comment (EdwinGuzman) [reply packages/opds-web-client/src/components/__tests__/DownloadButton-test.tsx]: Yes, let's keep `EPUB` for now.

**Comment by EdwinGuzman** [comment]
This looks good. I think the fulfillment links are slightly different but that doens't have to be in this PR.

File: packages/opds-web-client/src/components/__tests__/DownloadButton-test.tsx (hunk 5 above)
Comment (kristojorg) [reply packages/opds-web-client/src/components/__tests__/DownloadButton-test.tsx]: Switched it back : )

**Comment by kristojorg** [comment]
I don't have permission to do the actual merge, so I'll let y'all do that when you're ready.

**Comment by kristojorg** [comment]
Ready to merge this?

**Comment by EdwinGuzman** [comment]
Yes, merging now. Thanks.
